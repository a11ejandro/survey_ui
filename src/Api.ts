import axios, { AxiosResponse } from 'axios'
import { Statistics, Task } from './module'

const socketAddress = process.env.API_URL || ''
const axiosInstance = axios.create({
  baseURL: socketAddress + '/api/v1',
  timeout: 10000,
  headers: {
    common: {
      'Content-Type': 'application/json'
    }
  }
})


export async function createTask (task:Task) : Promise<AxiosResponse> {
  try {
    return await axiosInstance.post('tasks', task)
  } catch (e) {
    console.log(e)
    return Promise.reject(e)
  }
}

export async function getStatistics () : Promise<Statistics[]> {
  try {
    const response = await axiosInstance.get('statistics')
    return response.data as Statistics[]
  } catch (e) {
    console.log(e)
    return Promise.reject<Statistics[]>(e)
  }
}

