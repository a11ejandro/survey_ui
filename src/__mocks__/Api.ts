import { Statistics, Task } from '../module'

const promisify = (resolver: any) => {
  return new Promise((resolve, reject) => {
    resolve(resolver)
  })
}

export const statistics = [
  { collection_size: 1000, duration: 0.02, handler_type: 'go' },
  { collection_size: 1000, duration: 0.05, handler_type: 'ruby no AR'},
  { collection_size: 1000, duration: 0.1, handler_type: 'ruby + AR'}
]

export function getStatistics () {
  return promisify(statistics as Statistics[])
}

export function createTask (task: Task) {
  return promisify(true)
}
