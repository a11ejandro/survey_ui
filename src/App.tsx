import * as React from 'react'
import Paper from '@material-ui/core/Paper'
import Grid from '@material-ui/core/Grid'
import Button from '@material-ui/core/Button'

import { Statistics, Task } from './module'
import { createTask, getStatistics } from './Api'
import './App.css'
import StatisticsChart from './StatisticsChart'
import TaskForm from './TaskForm'

interface State {
  results: Statistics[]
}

class App extends React.Component<{}, State> {
  private timeout: number

  constructor(props: {}) {
    super(props)
    this.state = {
      results: []
    }

    this.createTask = this.createTask.bind(this)
    this.getResults = this.getResults.bind(this)
  }

  public componentWillUnmount () {
    clearTimeout(this.timeout)
  }

  public render() {
    return (
      <div className="App">
        <Paper>
          <Grid container spacing={24}>
            <Grid item xs={12}>
              <h3>Client for survey calculator</h3>
            </Grid>
            <Grid item xs={12}>
              <TaskForm createdCallback={this.createTask}/>
            </Grid>
            <Grid item xs={12}>
              <Button
                variant="outlined"
                color="primary"
                onClick={this.getResults}>
                Get last results
              </Button>
            </Grid>
            
            <Grid item xs={12}>
              <StatisticsChart data={this.state.results}/>
            </Grid>
          </Grid>
        </Paper>
      </div>
    )
  }

  private createTask (task: Task) {
    createTask(task).then((success: boolean) => {
      setTimeout(this.getResults, 3000)
    })
  }

  private getResults () {
    getStatistics().then((data: Statistics[]) => {
      this.setState({
        results: data
      })
    })
  }
}

export default App
