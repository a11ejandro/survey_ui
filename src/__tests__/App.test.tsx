import * as React from 'react'
import toJson from 'enzyme-to-json'
import { shallow } from 'enzyme'

import App from '../App'
import * as Api from '../Api'

jest.mock('../Api')
jest.useFakeTimers()

let mountedApp: any
const app = () => {
  if (!mountedApp) {
    mountedApp = shallow(<App/>)
  }
  return mountedApp
}

it ('renders correctly', () => {
  expect(toJson(app())).toMatchSnapshot()
})

it ('clears timeout on unmount', () => {
  app().instance().componentWillUnmount()
  expect(clearTimeout).toHaveBeenCalledTimes(1)
})

describe ('results', () => {
  it ('is initialized as empty array', () => {
    expect(app().state().results).toHaveLength(0)
  })

  it ('gets results from Api', async () => {
    const spyOnGetStatistics = jest.spyOn(Api, 'getStatistics')
    await app().instance().getResults()
    expect(spyOnGetStatistics).toBeCalled()
    expect(app().instance().state.results).toHaveLength(3)

    spyOnGetStatistics.mockReset()
    spyOnGetStatistics.mockRestore()
  })
})

describe ('task creation', () => {
  const newTask = {
    page: 1,
    per_page: 1000
  }

  it ('posts results to Api', async () => {
    const spyOnCreateTask = jest.spyOn(Api, 'createTask')
    await app().instance().createTask(newTask)
    expect(spyOnCreateTask).toBeCalled()

    spyOnCreateTask.mockReset()
    spyOnCreateTask.mockRestore()
  })

  it ('delays result refresh', async () => {
    await app().instance().createTask(newTask)
    expect(setTimeout).toHaveBeenLastCalledWith(app().instance().getResults, 3000)
  })

  describe ('toast notifications', () => {
    beforeEach(() => {
      app().setState({toasts: []})
    })

    it ('displays success notification', async () => {
      await app().instance().createTask(newTask)
      expect(app().state().toasts).toHaveLength(1)
      const toasts = app().find('#toast-container').children()
      expect(toasts).toHaveLength(1)
      expect(toasts.first().props().type).toBe('success')
    })

    it ('displays error', async () => {
      const spyOnCreateTask = jest.spyOn(Api, 'createTask').mockImplementation(() => Promise.reject)

      await app().instance().createTask(newTask)
      expect(app().state().toasts).toHaveLength(1)
      const toasts = app().find('#toast-container').children()
      expect(toasts).toHaveLength(1)
      expect(toasts.first().props().type).toBe('warning')

      spyOnCreateTask.mockReset()
      spyOnCreateTask.mockRestore()
    })

    it ('removes all toasts on click', async () => {
      await app().instance().createTask(newTask)
      await app().instance().createTask(newTask)
      expect(app().state().toasts).toHaveLength(2)
      const firstToast = app().find('#toast-container').children().first()
      firstToast.simulate('click')
      expect(app().state().toasts).toHaveLength(0)
      const toasts = app().find('#toast-container').children()
      expect(toasts).toHaveLength(0)
    })
  })
})
