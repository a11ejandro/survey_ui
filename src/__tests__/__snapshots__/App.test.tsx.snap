// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="App"
>
  <WithStyles(Paper)>
    <WithStyles(Grid)
      container={true}
      spacing={24}
    >
      <WithStyles(Grid)
        item={true}
        xs={12}
      >
        <h3>
          Client for survey calculator
        </h3>
      </WithStyles(Grid)>
      <WithStyles(Grid)
        item={true}
        xs={12}
      >
        <TaskForm
          createdCallback={[Function]}
        />
      </WithStyles(Grid)>
      <WithStyles(Grid)
        item={true}
        xs={12}
      >
        <WithStyles(Button)
          color="primary"
          onClick={[Function]}
          variant="outlined"
        >
          Get last results
        </WithStyles(Button)>
      </WithStyles(Grid)>
      <WithStyles(Grid)
        item={true}
        xs={12}
      >
        <StatisticsChart
          data={Array []}
        />
      </WithStyles(Grid)>
    </WithStyles(Grid)>
  </WithStyles(Paper)>
</div>
`;
